File: src/main/ele_action/handlers/file_render/adapter/ShadowAdapter.js
// src/main/ele_action/handlers/pdf/adapter/ShadowAdapter.js
import { IAdapter } from './IAdapter';

export class ShadowAdapter extends IAdapter {
  constructor(config) {
    super();
    this.config = config;
    this.pages = [];
    this.currentPageIndex = -1;
    this.currentPage = null;

    const [width, height] = [config.pageWidth, config.pageHeight];
    this.pageWidth = Math.ceil(config.landscape ? height : width);
    this.pageHeight = Math.ceil(config.landscape ? width : height);

    // 创建第一页
    this.createNewPage();
  }

  createNewPage() {
    this.currentPageIndex++;

    this.currentPage = {
      index: this.currentPageIndex,
      width: this.pageWidth,
      height: this.pageHeight,
      elements: []
    };

    this.pages.push(this.currentPage);
  }

  addPage() {
    this.createNewPage();
  }

  saveState() {
    this.currentPage.elements.push({
      type: 'saveState',
      timestamp: Date.now()
    });
  }

  restoreState() {
    this.currentPage.elements.push({
      type: 'restoreState',
      timestamp: Date.now()
    });
  }

  setTransform({ a, b, c, d, e, f }) {
    this.currentPage.elements.push({
      type: 'transform',
      matrix: { a, b, c, d, e, f },
      timestamp: Date.now()
    });
  }

  drawText({ text, x, y, size = 12 }) {
    this.currentPage.elements.push({
      type: 'text',
      text,
      x,
      y,
      fontSize: size,
      timestamp: Date.now()
    });
  }

  setLineStyle({ width, color }) {
    this.currentLineWidth = width;
    this.currentLineColor = color;

    this.currentPage.elements.push({
      type: 'lineStyle',
      width,
      color,
      timestamp: Date.now()
    });
  }

  drawLine({ x1, y1, x2, y2, dash }) {
    this.currentPage.elements.push({
      type: 'line',
      x1,
      y1,
      x2,
      y2,
      width: this.currentLineWidth || 1,
      color: this.currentLineColor || '#000000',
      dashed: !!dash,
      dashPattern: dash || null,
      timestamp: Date.now()
    });
  }

  fillRect({ x, y, width, height, color }) {
    this.currentPage.elements.push({
      type: 'rect',
      x,
      y,
      width,
      height,
      color,
      timestamp: Date.now()
    });
  }

  async drawImage({ data, x, y, width, height, rotation = 0 }) {
    // 调整坐标使其表示旋转后的视觉位置
    let adjustedX = x;
    let adjustedY = y;

    if (rotation === 180) {
      adjustedX = x - width;
      adjustedY = y + height;
    }

    this.currentPage.elements.push({
      type: 'image',
      x: adjustedX,
      y: adjustedY,
      width,
      height,
      rotation,
      rotated: rotation !== 0,
      dataType: data.base64 ? 'base64' : (data.path ? 'path' : 'unknown'),
      dataPath: data.path || null,
      dataSize: data.base64 ? data.base64.length : null,
      timestamp: Date.now()
    });
  }


  getPageSize() {
    return {
      width: this.pageWidth,
      height: this.pageHeight
    };
  }

  async finalize() {
    // 返回完整的绘制记录
    return {
      config: {
        pageSize: `${this.pageWidth}x${this.pageHeight}`,
        landscape: this.config.landscape,
        compressLevel: this.config.compressLevel
      },
      totalPages: this.pages.length,
      pages: this.pages.map(page => ({
        index: page.index,
        width: page.width,
        height: page.height,
        elementCount: page.elements.length,
        elements: page.elements
      })),
      summary: this.generateSummary()
    };
  }

  generateSummary() {
    const summary = {
      totalElements: 0,
      byType: {},
      byPage: []
    };

    this.pages.forEach(page => {
      const pageStats = {
        pageIndex: page.index,
        elements: {}
      };

      page.elements.forEach(element => {
        summary.totalElements++;

        // 统计类型
        if (!summary.byType[element.type]) {
          summary.byType[element.type] = 0;
        }
        summary.byType[element.type]++;

        // 统计每页
        if (!pageStats.elements[element.type]) {
          pageStats.elements[element.type] = 0;
        }
        pageStats.elements[element.type]++;
      });

      summary.byPage.push(pageStats);
    });

    return summary;
  }

  // 测试辅助方法
  getElementsOfType(type, pageIndex = null) {
    if (pageIndex !== null) {
      return this.pages[pageIndex]?.elements.filter(e => e.type === type) || [];
    }

    return this.pages.flatMap(page =>
      page.elements.filter(e => e.type === type)
    );
  }

  getTexts(pageIndex = null) {
    return this.getElementsOfType('text', pageIndex);
  }

  getLines(pageIndex = null) {
    return this.getElementsOfType('line', pageIndex);
  }

  getImages(pageIndex = null) {
    return this.getElementsOfType('image', pageIndex);
  }

  getRects(pageIndex = null) {
    return this.getElementsOfType('rect', pageIndex);
  }

  getDashedLines(pageIndex = null) {
    return this.getLines(pageIndex).filter(line => line.dashed);
  }

  getRotatedImages(pageIndex = null) {
    return this.getImages(pageIndex).filter(img => img.rotated);
  }

  findTextByContent(searchText, pageIndex = null) {
    return this.getTexts(pageIndex).filter(text =>
      text.text.includes(searchText)
    );
  }

  findElementsAtPosition(x, y, tolerance = 1, pageIndex = null) {
    const pages = pageIndex !== null ? [this.pages[pageIndex]] : this.pages;
    const results = [];

    pages.forEach(page => {
      page.elements.forEach(element => {
        let matches = false;

        if (element.type === 'text' || element.type === 'image' || element.type === 'rect') {
          matches = Math.abs(element.x - x) <= tolerance &&
            Math.abs(element.y - y) <= tolerance;
        } else if (element.type === 'line') {
          matches = (Math.abs(element.x1 - x) <= tolerance && Math.abs(element.y1 - y) <= tolerance) ||
            (Math.abs(element.x2 - x) <= tolerance && Math.abs(element.y2 - y) <= tolerance);
        }

        if (matches) {
          results.push({ ...element, pageIndex: page.index });
        }
      });
    });

    return results;
  }
}

File: server/file_render/adapter/shadow_adapter.go
package adapter

import (
    "time"
)

// Element 绘制元素
type Element struct {
    Type      string                 `json:"type"`
    Timestamp int64                  `json:"timestamp"`
    Data      map[string]interface{} `json:"data,omitempty"`
}

// Page 页面
type Page struct {
    Index        int       `json:"index"`
    Width        int       `json:"width"`
    Height       int       `json:"height"`
    Elements     []Element `json:"elements"`
    ElementCount int       `json:"elementCount"`
}

// ShadowAdapter 影子适配器（用于调试和测试）
type ShadowAdapter struct {
    config           map[string]interface{}
    pages            []Page
    currentPageIndex int
    currentPage      *Page
    pageWidth        int
    pageHeight       int
    currentLineWidth float64
    currentLineColor string
}

// NewShadowAdapter 创建新的影子适配器
func NewShadowAdapter(config map[string]interface{}) *ShadowAdapter {
    landscape, _ := config["landscape"].(bool)
    pageWidth, _ := config["pageWidth"].(float64)
    pageHeight, _ := config["pageHeight"].(float64)

    width := int(pageWidth)
    height := int(pageHeight)

    if landscape {
        width, height = height, width
    }

    adapter := &ShadowAdapter{
        config:           config,
        pages:            []Page{},
        currentPageIndex: -1,
        pageWidth:        width,
        pageHeight:       height,
    }

    adapter.createNewPage()
    return adapter
}

// createNewPage 创建新页面
func (s *ShadowAdapter) createNewPage() {
    s.currentPageIndex++
    s.currentPage = &Page{
        Index:    s.currentPageIndex,
        Width:    s.pageWidth,
        Height:   s.pageHeight,
        Elements: []Element{},
    }
    s.pages = append(s.pages, *s.currentPage)
}

// AddPage 添加新页面
func (s *ShadowAdapter) AddPage() error {
    s.createNewPage()
    return nil
}

// SaveState 保存状态
func (s *ShadowAdapter) SaveState() error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "saveState",
        Timestamp: time.Now().UnixMilli(),
    })
    return nil
}

// RestoreState 恢复状态
func (s *ShadowAdapter) RestoreState() error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "restoreState",
        Timestamp: time.Now().UnixMilli(),
    })
    return nil
}

// SetTransform 设置变换矩阵
func (s *ShadowAdapter) SetTransform(transform Transform) error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "transform",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "matrix": transform,
        },
    })
    return nil
}

// DrawText 绘制文本
func (s *ShadowAdapter) DrawText(options TextOptions) error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "text",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "text":     options.Text,
            "x":        options.X,
            "y":        options.Y,
            "fontSize": options.Size,
        },
    })
    return nil
}

// SetLineStyle 设置线条样式
func (s *ShadowAdapter) SetLineStyle(style LineStyle) error {
    s.currentLineWidth = style.Width
    s.currentLineColor = style.Color

    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "lineStyle",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "width": style.Width,
            "color": style.Color,
        },
    })
    return nil
}

// DrawLine 绘制线条
func (s *ShadowAdapter) DrawLine(options LineOptions) error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "line",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "x1":          options.X1,
            "y1":          options.Y1,
            "x2":          options.X2,
            "y2":          options.Y2,
            "width":       s.currentLineWidth,
            "color":       s.currentLineColor,
            "dashed":      len(options.Dash) > 0,
            "dashPattern": options.Dash,
        },
    })
    return nil
}

// FillRect 填充矩形
func (s *ShadowAdapter) FillRect(options RectOptions) error {
    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "rect",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "x":      options.X,
            "y":      options.Y,
            "width":  options.Width,
            "height": options.Height,
            "color":  options.Color,
        },
    })
    return nil
}

// DrawImage 绘制图片
func (s *ShadowAdapter) DrawImage(options ImageOptions) error {
    adjustedX := options.X
    adjustedY := options.Y

    if options.Rotation == 180 {
        adjustedX = options.X - options.Width
        adjustedY = options.Y + options.Height
    }

    dataType := "unknown"
    if options.Data.Base64 != "" {
        dataType = "base64"
    } else if options.Data.Path != "" {
        dataType = "path"
    }

    s.currentPage.Elements = append(s.currentPage.Elements, Element{
        Type:      "image",
        Timestamp: time.Now().UnixMilli(),
        Data: map[string]interface{}{
            "x":        adjustedX,
            "y":        adjustedY,
            "width":    options.Width,
            "height":   options.Height,
            "rotation": options.Rotation,
            "rotated":  options.Rotation != 0,
            "dataType": dataType,
            "dataPath": options.Data.Path,
            "dataSize": len(options.Data.Base64),
        },
    })
    return nil
}

// GetPageSize 获取页面尺寸
func (s *ShadowAdapter) GetPageSize() PageSize {
    return PageSize{
        Width:  float64(s.pageWidth),
        Height: float64(s.pageHeight),
    }
}

// Finalize 完成并返回结果
func (s *ShadowAdapter) Finalize() (interface{}, error) {
    // 更新元素计数
    for i := range s.pages {
        s.pages[i].ElementCount = len(s.pages[i].Elements)
    }

    result := map[string]interface{}{
        "config": map[string]interface{}{
            "pageSize":      s.pageWidth,
            "landscape":     s.config["landscape"],
            "compressLevel": s.config["compressLevel"],
        },
        "totalPages": len(s.pages),
        "pages":      s.pages,
        "summary":    s.generateSummary(),
    }

    return result, nil
}

// generateSummary 生成摘要
func (s *ShadowAdapter) generateSummary() map[string]interface{} {
    summary := map[string]interface{}{
        "totalElements": 0,
        "byType":        make(map[string]int),
        "byPage":        []map[string]interface{}{},
    }

    totalElements := 0

    for _, page := range s.pages {
        pageStats := map[string]interface{}{
            "pageIndex": page.Index,
            "elements":  make(map[string]int),
        }

        for _, element := range page.Elements {
            totalElements++

            // 统计类型
            byType := summary["byType"].(map[string]int)
            byType[element.Type]++

            // 统计每页
            elements := pageStats["elements"].(map[string]int)
            elements[element.Type]++
        }

        summary["byPage"] = append(summary["byPage"].([]map[string]interface{}), pageStats)
    }

    summary["totalElements"] = totalElements

    return summary
}
